### PR TITLE
feat: iCake

### PR DESCRIPTION
**Pull request by ChefMomota**
Before Release
- [x]  Medium link

### CODE REVIEW

File: src/state/pools/fetchIfoUserCredit.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import BigNumber from 'bignumber.js'
+import { getICakeddress } from 'utils/addressHelpers'
+import iCakeAbi from 'config/abi/iCake.json'
+import { multicallv2 } from 'utils/multicall'
+
+const fetchIfoUserCredit = async (account: string) => {
+  try {
+    const calls = [
+      {
+        address: getICakeddress(),
```
Comment (memoyil) [inline src/state/pools/fetchIfoUserCredit.ts:10]: Typo on method name

File: src/state/pools/index.ts [hunk 2]
```diff
@@ -298,6 +305,17 @@ export const fetchCakeVaultUserData = createAsyncThunk<SerializedLockedVaultUser
   },
 )
 
+export const fetchIfoUserCreditDataAsync =
+  (account: string): AppThunk =>
+  async (dispatch) => {
+    try {
+      const iCakeUserData = await fetchIfoUserCredit(account)
+      dispatch(setIfoUserCreditData(iCakeUserData))
```
Comment (memoyil) [inline src/state/pools/index.ts:313]: Imo No need to dispatch again, we can do this operation inside the slice directly

Comment (Chef-Yogi) [reply src/state/pools/index.ts]: @memoyil 
may I ask why you suggest we don't need to dispatch here, cound't find the difference between the pattern like "fetchPoolsUserDataAsync"?
thanks for your kindly help🙏

Comment (Chef-Yogi) [reply src/state/pools/index.ts]: @memoyil 
may I ask why you suggest we don't need to dispatch here, cound't find the difference between the pattern like "fetchPoolsUserDataAsync"?
thanks for your kindly help🙏

Comment (ChefMomota) [reply src/state/pools/index.ts]: Hey @memoyil , changed the state. 
But not sure what you mean in here.Or do we have any example?

Comment (memoyil) [reply src/state/pools/index.ts]: You can create a reducer in extraReducers of poolsSlice for fetchIfoUserCreditDataAsync.fulfilled then change the state inside that reducer. Therefore we will not need to call and start the dispatch process again

Comment (ChefMomota) [reply src/state/pools/index.ts]: @memoyil  Thanks.

File: src/views/Ifos/index.tsx [hunk 3]
```diff
@@ -2,15 +2,19 @@ import { SubMenuItems } from '@pancakeswap/uikit'
...
 import { useRouter } from 'next/router'
+import { usePoolsPageFetch, useFetchIfo } from 'state/pools/hooks'
 import Hero from './components/Hero'
+import IfoProvider from './contexts/IfoContext'
 
 export const IfoPageLayout = ({ children }) => {
   const { t } = useTranslation()
   const router = useRouter()
   const isExact = router.route === '/ifo'
+  usePoolsPageFetch()
```
Comment (chef-jojo) [inline src/views/Ifos/index.tsx:13]: Is this too much to fetch all pools for ifo page?

Comment (ChefMomota) [reply src/views/Ifos/index.tsx]: @chef-jojo  Updated.

File: src/state/pools/fetchUserIfo.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        address: getICakeAddress(),
+        name: 'getUserCredit',
+        params: [account],
+      },
+    ]
+
+    const [creditResponse] = await multicallv2(iCakeAbi, calls)
+    return new BigNumber(creditResponse.toString()).toJSON()
+  } catch (error) {
+    return null
```
Comment (chef-jojo) [inline src/state/pools/fetchUserIfo.ts:35]: unnecessary multicall

File: src/state/pools/selectors.ts [hunk 5]
```diff
@@ -42,3 +46,11 @@ export const poolsWithVaultSelector = createSelector(
     return { pools: [cakeAutoVaultWithApr, ...withoutCakePool], userDataLoaded }
   },
 )
+
+export const ifoCreditSelector = createSelector([selectIfoUserCredit], (ifoUserCredit) => {
+  return new BigNumber(ifoUserCredit)
+})
+
+export const ifoCeilingSelector = createSelector([selectIfo], (ifoData) => {
+  return new BigNumber(ifoData.ceiling)
```
Comment (chef-jojo) [inline src/state/pools/selectors.ts:55]: maybe null

File: src/views/Ifos/components/IfoVesting/LockCakeButton.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import React from 'react'
```
Comment (chef-jojo) [inline src/views/Ifos/components/IfoVesting/LockCakeButton.tsx:1]: no need import React

File: src/views/Ifos/components/IfoFoldableCard/IfoPoolCard/IfoVestingCard/ProgressStepper.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+const ProgressStepper: React.FC<CountdownProps> = ({ steps, activeStepIndex }) => {
+  const { t } = useTranslation()
+  return (
+    <Flex>
+      {steps.map((step, index) => {
+        const isPastSpacer = index < activeStepIndex
+        const stepText = t(step.text).toUpperCase()
+
+        return (
+          <Fragment key={_uniqueId('Ifo-Vesting-Step-')}>
```
Comment (chef-jojo) [inline src/views/Ifos/components/IfoFoldableCard/IfoPoolCard/IfoVestingCard/ProgressStepper.tsx:27]: use index is fine here

**Comment by chef-jojo** [comment]
linting

File: src/state/pools/hooks.ts [hunk 8]
```diff
@@ -86,6 +90,29 @@ export const usePoolsPageFetch = () => {
...
+      if (account) {
+        dispatch(fetchPoolsUserDataAsync(account))
+        dispatch(fetchCakeVaultUserData({ account }))
+        dispatch(fetchUserIfoCreditDataAsync(account))
+      }
+    })
+  }, [dispatch, account])
+
+  useEffect(() => {
+    batch(() => {
```
Comment (memoyil) [inline src/state/pools/hooks.ts:110]: Batch can be removed here

File: src/state/pools/index.ts [hunk 9]
```diff
@@ -298,6 +305,24 @@ export const fetchCakeVaultUserData = createAsyncThunk<SerializedLockedVaultUser
   },
 )
 
+export const fetchIfoPublicDataAsync = () => async (dispatch) => {
```
Comment (chefilip) [inline src/state/pools/index.ts:308]: This function should be thunk action?

File: src/views/Ifos/components/IfoFoldableCard/IfoPoolCard/IfoVestingCard/ProgressStepper.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    isPastSpacer ? theme.colors.textSubtle : theme.colors.textDisabled};
+`
+
+const ProgressStepper: React.FC<CountdownProps> = ({ steps, activeStepIndex }) => {
+  const { t } = useTranslation()
+  return (
+    <Flex>
+      {steps.map((step, index: number) => {
+        const isPastSpacer = index < activeStepIndex
+        const stepText = t(step.text).toUpperCase()
```
Comment (chefilip) [inline src/views/Ifos/components/IfoFoldableCard/IfoPoolCard/IfoVestingCard/ProgressStepper.tsx:23]: I don't think this translation would work.

File: src/state/pools/hooks.ts (hunk 8 above)
Comment (ChefMomota) [reply src/state/pools/hooks.ts:110]: `usePoolsPageFetch`  need to remove `batch` ?

https://github.com/ChefMomota/pancake-frontend/blob/bf14aa10ca7ba19b2ce27224a2a9737e30b2ee12/src/state/pools/hooks.ts#L71-L91

File: src/views/Ifos/components/IfoFoldableCard/IfoPoolCard/IfoVestingCard/ProgressStepper.tsx (hunk 10 above)
Comment (ChefMomota) [reply src/views/Ifos/components/IfoFoldableCard/IfoPoolCard/IfoVestingCard/ProgressStepper.tsx:23]: Why you think that translation doesn't work? 
Same as Trading Competition component.

https://github.com/pancakeswap/pancake-frontend/blob/2dfb6ffc860738534de3877d7df00d5bfea980e9/src/views/TradingCompetition/components/Countdown/ProgressStepper.tsx#L22-L39

File: src/state/pools/index.ts (hunk 9 above)
Comment (ChefMomota) [reply src/state/pools/index.ts]: @chefilip  done

File: src/state/pools/hooks.ts (hunk 8 above)
Comment (memoyil) [reply src/state/pools/hooks.ts:110]: Yes, batch is only needed when there is multiple dispatches to reduce renders

**Review by chef-jojo — APPROVED** [review]

**Pull request merged**